### PR TITLE
feat: auto concurrency limiter of adaptive service 

### LINKED ISSUE #1834: GSoC 2022: Explore a new generation of flexible loadbalance algorithms
Flexible loadbalance is an intelligent cluster scheduling algorithm, which can dynamically allocate traffic according to real-time machine load and hardware configuration, so as to maintain the best performance of the whole system. We will set up a flexible service performance test environment to simulate the real cluster status, including instance capacity expansion, timeout, offline due to failure, etc.

So if you are interested in this project and the objective described in the proposal, please leave comments below so we can further exchange opinions.

Please don't forget to apply for GSoC from its official website when the application slot opens.

### PR DESCRIPTION
**Pull request by CoolIceV**
**What this PR does**: 
Provides an auto concurrency limiter in unstable environments.
This is a component of adaptive service at the provider side.
The component at the consumer side is in #2067.

**Which issue(s) this PR fixes**: 
Fixes #1834 

**You should pay attention to items below to ensure your pr passes our ci test**
We do not merge pr with ci tests failed

- [ ] All ut passed (run 'go test ./...' in project root)
- [ ] After go-fmt ed , run 'go fmt project' using goland.
- [ ] Golangci-lint passed, run 'sudo golangci-lint run' in project root.
- [ ] Your new-created file needs to have [apache license](https://raw.githubusercontent.com/dubbogo/resources/master/tools/license/license.txt) at the top, like other existed file does.
- [ ] All integration test passed. You can run integration test locally (with docker env). Clone our [dubbo-go-samples](https://github.com/apache/dubbo-go-samples) project and replace the go.mod to your dubbo-go, and run 'sudo sh start_integration_test.sh' at root of samples project root. (M1 Slice is not Support)

### CODE REVIEW

**Comment by CoolIceV** [comment]
This concurrency limiter uses CPU load as a limit switch to automatically calculate the capacity of the service when the load is too high, the effect is as follows.

image

QPS increases gradually over time. When the CPU load is high, flow limiting is automatically turned on. Then successful requests and CPU load are maintained at a relatively stable level.

File: filter/adaptivesvc/limiter/auto_concurrency_limiter.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	MaxSampleCount     = 500
+	CpuDecay           = 0.95
+)
+
+type AutoConcurrency struct {
+	sync.RWMutex
+
+	exploreRatio         float64
+	emaFactor            float64
+	noLoadLatency        float64 //duration
```
Comment (justxuewei) [inline filter/adaptivesvc/limiter/auto_concurrency_limiter.go:58]: nit: //duration -> // duration

File: filter/adaptivesvc/limiter/auto_concurrency_limiter.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	maxQPS               float64
+	halfSampleIntervalMS int64
+	maxConcurrency       uint64
+
+	// metrics of the current round
+	startSampleTimeUs  int64
+	lastSamplingTimeUs *atomic.Int64
+	resetLatencyUs     int64 // time to reset noLoadLatency
+	remeasureStartUs   int64 //time to reset req data (sampleCount, totalSampleUs, totalReqCount)
```
Comment (justxuewei) [inline filter/adaptivesvc/limiter/auto_concurrency_limiter.go:67]: nit: //time -> // time

File: filter/adaptivesvc/limiter/auto_concurrency_limiter.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	go cpuproc()
+}
+
+// cpu = cpuᵗ⁻¹ * decay + cpuᵗ * (1 - decay)
+func cpuproc() {
+	ticker := time.NewTicker(time.Millisecond * 500) // same to cpu sample rate
+	defer func() {
+		ticker.Stop()
+		if err := recover(); err != nil {
+			go cpuproc()
```
Comment (justxuewei) [inline filter/adaptivesvc/limiter/auto_concurrency_limiter.go:85]: Add a warning log to let users know what is happening here.

File: filter/adaptivesvc/limiter/auto_concurrency_limiter.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+)
+
+// These parameters may need to be different between services
+const (
+	MaxExploreRatio    = 0.3
+	MinExploreRatio    = 0.06
+	SampleWindowSizeMs = 1000
+	MinSampleCount     = 40
+	MaxSampleCount     = 500
+	CpuDecay           = 0.95
```
Comment (justxuewei) [inline filter/adaptivesvc/limiter/auto_concurrency_limiter.go:50]: nit: CpuDecay -> CPUDecay

File: filter/adaptivesvc/limiter/auto_concurrency_limiter.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+	for range ticker.C {
+		usage := cpu.CpuUsage()
+		prevCPU := cpuLoad.Load()
+		curCPU := uint64(float64(prevCPU)*CpuDecay + float64(usage)*(1.0-CpuDecay))
+		logger.Debugf("current cpu usage: %d", curCPU)
+		cpuLoad.Store(curCPU)
+	}
+}
+
+func CpuUsage() uint64 {
```
Comment (justxuewei) [inline filter/adaptivesvc/limiter/auto_concurrency_limiter.go:98]: CpuUsage -> CPUUsage

**Comment by justxuewei** [comment]
Merged into `3.0-adaptive-stream` branch to make the running benchmark more convenient. Once the current version is in line with expectations, it will be merged into `3.0` branch.

**Pull request merged**